File: stealthwall/middleware/fastapi/stealthwall_fastapi/middleware.py
"""STEALTHWALL — FastAPI/Starlette middleware (pure ASGI, PyPI target).

FastAPI ONLY (plan Section 5). Django and Flask are explicitly unsupported.

Flow per request:
  PRE-REQUEST enforcement — active blocks -> 403; captcha-required ->
  403 + mCaptcha widget payload; throttled -> 429. Enforcement happens
  BEFORE the app runs so blocked sources never touch business logic.
  POST-RESPONSE scoring — the observed request+response event is appended
  to the per-IP sliding window, features extracted per
  docs/feature_extraction_spec.md, blended through the scoring pipeline,
  and handed to the graduated-response engine.

ASN tags are attached ONLY to dashboard feed entries, never to any
client-facing response (info-leak caveat, plan Section 6).
"""

from __future__ import annotations

import asyncio
import collections
import json
import sys
import threading
import time
from pathlib import Path
from typing import Deque, Dict, Optional

_ROOT = Path(__file__).resolve().parents[3]
for p in (str(_ROOT), str(_ROOT / "block_engine")):
    if p not in sys.path:
        sys.path.insert(0, p)

from config.defaults import DASHBOARD_FEED_POLL_SECONDS  # noqa: E402
from stealthwall_fastapi.features import extract_features  # noqa: E402
from stealthwall_fastapi.scoring import ScoringPipeline  # noqa: E402


class DecisionFeed:
    """In-memory ring buffer for the dashboard; thread-safe."""

    def __init__(self, maxlen: int = 500):
        self._buf = collections.deque(maxlen=maxlen)
        self._lock = threading.Lock()

    def add(self, entry: dict) -> None:
        with self._lock:
            self._buf.append(entry)

    def recent(self, n: int = 50) -> list:
        with self._lock:
            items = list(self._buf)[-n:]
        return list(reversed(items))


class StealthWallMiddleware:
    """Wrap an ASGI app:

        app.add_middleware(StealthWallMiddleware,
                           scorer=..., response_engine=...)
    """

    def __init__(
        self,
        app,
        scorer: Optional[ScoringPipeline] = None,
        response_engine=None,          # GraduatedResponseEngine-compatible
        blocker=None,                  # BlockWriterClient/DryRun compatible
        adaptive_layer=None,           # models/adaptive_scoring instance
        cf_pusher=None,
        observe_only: bool = False,
        trusted_proxy_hops: int = 0,
        feed: Optional[DecisionFeed] = None,
        captcha_provider=None,
        exclude_paths: Optional[List[str]] = None,
    ):
        self.app = app
        self.scorer = scorer
        self.response_engine = response_engine
        self.blocker = blocker
        self.adaptive_layer = adaptive_layer
        self.cf_pusher = cf_pusher
        self.observe_only = observe_only
        self.trusted_proxy_hops = trusted_proxy_hops
        self.feed = feed or DecisionFeed()
        self.captcha_provider = captcha_provider
        self.exclude_paths = set(exclude_paths or [])
        self.windows: Dict[str, Deque[dict]] = collections.defaultdict(
            lambda: collections.deque())
        self._captcha_required: Dict[str, float] = {}

    # ------------------------------------------------------------------ ASGI
    async def __call__(self, scope, receive, send):
        if scope["type"] != "http" or scope.get("path", "") in self.exclude_paths:
            await self.app(scope, receive, send)
            return

        ip = self._client_ip(scope)
        start = time.time()

        action = self._pre_enforce(ip)
        if action is not None:
            kind, payload = action
            await self._respond_json(send, **payload)
            self.feed.add({"ip": ip, "action": f"rejected:{kind}",
                           "at": start})
            return

        status_holder = {"status": 0}
        payload_holder = {"payload": ""}

        app_receive = receive
        method = scope.get("method", "GET").upper()
        if method in ("POST", "PUT", "PATCH", "DELETE"):
            try:
                buffered_messages = []
                body_bytes = b""
                more_body = True
                while more_body and len(body_bytes) < 4096:
                    msg = await receive()
                    buffered_messages.append(msg)
                    body_bytes += msg.get("body", b"")
                    more_body = msg.get("more_body", False)

                payload_holder["payload"] = body_bytes.decode("utf-8", "replace")[:512]

                async def replay_receive():
                    if buffered_messages:
                        return buffered_messages.pop(0)
                    return await receive()

                app_receive = replay_receive
            except Exception:
                app_receive = receive

        async def send_wrapper(message):
            if message["type"] == "http.response.start":
                status_holder["status"] = message["status"]
            await send(message)

        try:
            await self.app(scope, app_receive, send_wrapper)
        finally:
            await asyncio.to_thread(self._post_score, scope, ip,
                                    status_holder["status"], start, payload_holder["payload"])

    # ------------------------------------------------------------- client IP
    def _client_ip(self, scope) -> str:
        client = scope.get("client") or ("unknown", 0)
        ip = client[0]
        headers = {k.decode().lower(): v.decode() for k, v in
                   scope.get("headers", [])}
        if self.trusted_proxy_hops > 0:
            xff = headers.get("x-forwarded-for")
            if xff:
                hops = [h.strip() for h in xff.split(",")]
                idx = -self.trusted_proxy_hops - 1
                if abs(idx) <= len(hops):
                    return hops[idx]
        return ip

    # ------------------------------------------------------------ pre-request
    def _pre_enforce(self, ip: str):
        """Returns None to proceed, else (kind, response-spec)."""
        if self.blocker is not None and not self.observe_only:
            try:
                state = self.blocker.request(
                    {"op": "check", "ip": ip})
                if state.get("blocked"):
                    return "blocked", dict(status=403, body={
                        "error": "blocked",
                        "reason": "source temporarily blocked",
                    })
            except Exception as exc:  # noqa: BLE001 — fail-open on IPC loss
                self.feed.add({"ip": ip, "action": "ipc_error",
                               "detail": repr(exc), "at": time.time()})
        need_captcha_at = self._captcha_required.get(ip)
        if not self.observe_only and need_captcha_at \
                and time.time() < need_captcha_at:
            widget = {}
            if self.captcha_provider is not None:
                try:
                    widget = self.captcha_provider.widget_config()
                except Exception:  # noqa: BLE001
                    widget = {}
            return "captcha", dict(status=403, body={
                "error": "captcha_required", "widget": widget})
        return None

    # ----------------------------------------------------------- post-response
    def _post_score(self, scope, ip: str, status: int, start: float, payload: str = "") -> None:
        try:
            headers = {k.decode().lower(): v.decode("utf-8", "replace")
                       for k, v in scope.get("headers", [])}
            event = {
                "ts": round(start, 6),
                "method": scope.get("method", "GET").upper(),
                "path": scope.get("path", "/"),
                "status": status,
                "payload": payload or "",
                "headers": headers,
                "user_agent": headers.get("user-agent", ""),
                "is_auth_failure": bool(
                    status == 401 and scope.get("path") == "/login"),
            }
            window = self.windows[ip]
            window.append(event)
            vector = extract_features(list(window))
            if vector is None or self.scorer is None:
                return
            result = self.scorer.score(ip, vector)

            decision = None
            if self.response_engine is not None:
                decision = self.response_engine.decide_and_respond(
                    ip, result["final_score"])

            if decision is not None:
                if decision.action == "captcha":
                    ttl = 900  # challenge validity; widget retry inside TTL
                    self._captcha_required[ip] = time.time() + ttl
                entry = {"ip": ip,
                         **decision.to_dashboard_entry()}
                entry["raw_model_score"] = result["raw_model_score"]
                entry["latency_ms"] = round((time.time() - start) * 1000, 2)
                self.feed.add(entry)
        except Exception as exc:  # noqa: BLE001 — scoring must never break app
            self.feed.add({"ip": ip, "action": "scoring_error",
                           "detail": repr(exc), "at": time.time()})

    # ----------------------------------------------------------------- respond
    async def _respond_json(self, send, status: int, body: dict) -> None:
        payload = json.dumps(body).encode("utf-8")
        await send({
            "type": "http.response.start",
            "status": status,
            "headers": [(b"content-type", b"application/json")],
        })
        await send({"type": "http.response.body", "body": payload})

    # ------------------------------------------------------------- dashboard
    def mark_captcha_solved(self, ip: str) -> None:
        self._captcha_required.pop(ip, None)

    def dashboard_snapshot(self) -> dict:
        return {
            "feed": self.feed.recent(),
            "poll_seconds": DASHBOARD_FEED_POLL_SECONDS,
            "observe_only": self.observe_only,
        }

File: stealthwall/middleware/express/src/middleware.js
/**
 * STEALTHWALL — Express middleware (npm target).
 *
 * Mirrors stealthwall_fastapi/middleware.py:
 *  - PRE-REQUEST enforcement: active blocks -> 403; captcha-required ->
 *    403 + widget payload.
 *  - POST-RESPONSE scoring: append event to per-IP window, extract
 *    features, blend score, hand to the graduated-response engine.
 *
 * ASN tags live ONLY in dashboard feed entries, never client responses
 * (plan Section 6 info-leak caveat).
 */

'use strict';

const { extractFeatures } = require('./features');

class DecisionFeed {
  constructor(maxlen = 500) {
    this.maxlen = maxlen;
    this.buf = [];
  }
  add(entry) {
    this.buf.push(entry);
    if (this.buf.length > this.maxlen) this.buf.shift();
  }
  recent(n = 50) {
    return this.buf.slice(-n).reverse();
  }
}

function defaultAuthFailureCheck(status, reqPath) {
  return status === 401 && reqPath === '/login';
}

function stealthwall(options = {}) {
  const {
    scorer = null,
    responseEngine = null,
    blocker = null,
    observeOnly = false,
    feed = new DecisionFeed(),
    captchaProvider = null,
    isAuthFailure = defaultAuthFailureCheck,
  } = options;

  const windows = new Map(); // ip -> [events]
  const captchaRequired = new Map(); // ip -> expiry ts

  // ------------------------------------------------------------- post-score
  function postScore(req, res, startTs) {
    try {
      const ip = req.ip || (req.socket && req.socket.remoteAddress) || 'unknown';
      let win = windows.get(ip);
      if (!win) { win = []; windows.set(ip, win); }

      let payloadStr = '';
      if (req.body) {
        payloadStr = typeof req.body === 'string' ? req.body : JSON.stringify(req.body);
      } else if (req.rawBody) {
        payloadStr = String(req.rawBody);
      }
      if (payloadStr.length > 512) payloadStr = payloadStr.slice(0, 512);

      const event = {
        ts: startTs,
        method: (req.method || 'GET').toUpperCase(),
        path: req.originalUrl || req.url || '/',
        status: res.statusCode,
        payload: payloadStr,
        headers: (() => {
          const h = {};
          for (const [k, v] of Object.entries(req.headers || {})) {
            h[k.toLowerCase()] = Array.isArray(v) ? String(v[0]) : String(v);
          }
          return h;
        })(),
        user_agent: String((req.headers || {})['user-agent'] || ''),
        is_auth_failure: isAuthFailure(res.statusCode, req.path || ''),
      };
      win.push(event);

      if (!scorer) return;
      const vector = extractFeatures(win);
      if (!vector) return;
      const result = scorer.score(ip, vector);

      let decision = null;
      if (responseEngine) {
        decision = responseEngine.decideAndRespond(ip, result.final_score);
      }
      if (decision) {
        if (decision.action === 'captcha') {
          captchaRequired.set(ip, Date.now() / 1000 + 900);
        }
        feed.add({ ip, ...decision.toDashboardEntry(),
                   raw_model_score: result.raw_model_score });
      }
    } catch (err) {
      feed.add({ ip: 'unknown', action: 'scoring_error',
                 detail: String(err), at: Date.now() / 1000 });
    }
  }

  // ------------------------------------------------------------------ main
  return function stealthwallMiddleware(req, res, next) {
    const startTs = Date.now() / 1000;
    const ip = req.ip || (req.socket && req.socket.remoteAddress) || 'unknown';

    if (!observeOnly) {
      if (blocker && blocker.request) {
        try {
          const state = blocker.request({ op: 'check', ip });
          if (state.blocked) {
            feed.add({ ip, action: 'rejected:blocked', at: startTs });
            return res.status(403).json({
              error: 'blocked',
              reason: 'source temporarily blocked',
            });
          }
        } catch (err) {
          feed.add({ ip, action: 'ipc_error', detail: String(err),
                     at: startTs }); // fail-open on IPC loss, logged
        }
      }
      const needCaptcha = captchaRequired.get(ip);
      if (needCaptcha && Date.now() / 1000 < needCaptcha) {
        let widget = {};
        if (captchaProvider) {
          try { widget = captchaProvider.widgetConfig(); } catch (_) {}
        }
        feed.add({ ip, action: 'rejected:captcha', at: startTs });
        return res.status(403).json({ error: 'captcha_required', widget });
      }
    }

    res.on('finish', () => postScore(req, res, startTs));
    next();
  };
}

module.exports = { stealthwall, DecisionFeed };
